tooling/nargo_cli/tests/execution_success/trait_function_calls/src/main.nr

```diff
@@ -0,0 +1,188 @@
+// a more thorough test for trait function/method calls from within trait function/method calls
+
+// This tests all combinations of caller+callee pairs. For each of these, the following combination of properties are tried:
+//   *) method (has self parameter) vs function (no self parameter)
+//   *) default vs overriden vs overriden (no default)
+
+// TODO: function calls via `Self::` are not yet implemented
+// Once this is implemented, this test should be updated.
+
+// test order is:
+// 1) trait method -> trait method
+// 1a) trait default method -> trait default method
+// 1b) trait default method -> trait overriden method
+// 1c) trait default method -> trait overriden (no default) method
+// 1d) trait overriden method -> trait default method
+// 1e) trait overriden method -> trait overriden method
+// 1f) trait overriden method -> trait overriden (no default) method
+// 1g) trait overriden (no default) method -> trait default method
+// 1h) trait overriden (no default) method -> trait overriden method
+// 1i) trait overriden (no default) method -> trait overriden (no default) method
+// 2) trait method -> trait function
+// 2a) - subcases are the same as the above
+// ...
+// 2i)
+// 3{a..i}) trait function -> trait method
+// 4{a..i}) trait function -> trait function
```

Comment:
Are these the unimplemented cases mentioned by the TODO above?

Reply:
Yes, these are the unimplemented cases, mentioned in the TODO above. I know it's not always practical to test every combination of something, unless the number of test conditions is small, yielding a small number of combinations. In this case, we have 3*3=9 cases, that will be extended to 2*2*3*3=27 cases, once function calls via Self:: (a.k.a. static methods) are implemented. This is a very practical number of cases for an automated test and I'm not planning to increase the test conditions for this test any further, because it serves a very specific purpose - it tests whether function call resolution is done correctly with regards to choosing between trait default and trait impl ("override") implementation. This is also a feature that is specific to traits and is different, compared to e.g. struct impl function calls, where there's no such thing as default and override implementations (and combinations). We're not planning to adding any more dimensions to the test, that are unrelated to that (for example, once traits for non-struct types are implemented, we're *not* going to add every non-struct type as a new dimension to *this* test, but we'll definitely add a *new* test that just implements a trait for each type - that's because implementing traits for non-struct types is something that is completely independent from trait function call resolution).